File: werewolves/app/game.go
package werewolves

import (
  "bufio"
  "os"
  "sort"
  "fmt"
  "math/rand"
)

var reader = bufio.NewReader(os.Stdin)

func Play(xml GameXML) error {
  fmt.Printf("WELCOME TO %v\n", xml.Title)
  fmt.Print("WHAT IS YOUR NAME, EXPLORER?")
  name, err := reader.ReadString('\n')
  if err != nil {
    return err
  }
  game := buildGame(xml, name)
  //fmt.Println(game);
  for game.Player.CurrentRoom != game.Maze.Exit && game.Player.Strength > 0 {
    room, ok := game.Maze.Rooms[game.Player.CurrentRoom]
    if !ok {
      panic("rooms out of sync")
    }
    for _, line := range room.Description {
      fmt.Println(line)
    }
    fmt.Println(game.Player)
    if room.Contents > 0 {
      fmt.Println("THERE IS TREASURE HERE WORTH %v GOLD!")
      game.Player.Wealth += room.Contents
      room.Contents = 0
    } else if room.Contents < 0 {
      monster, ok := game.Monsters[room.Contents];
      if ok {
        combat(game.Player, monster)
      } else {
        room.Contents = 0
      }
    }
    if game.Player.Strength > 0 {
      fmt.Println("What do you want to do?")
    }
  }
  if game.Player.Strength <= 0 {
    fmt.Println("YOU HAVE DIED.........")
  }
  return nil
}

func combat(player *Player, monster *Monster) {
  fmt.Println("DANCER...THERE IS A MONSTER HERE....")
  fmt.Println("IT IS A %v")
  fmt.Println("THE DANGER LEVEL IS %v!!")
  first := true;
  list := ""
  for key, _ := range combatActions {
    if (first) {
      first = false
    } else {
      list += ","
    }
    list += key
  }
  repeat := true;
  for repeat {
    action := func(player *Player, monster *Monster) bool {
      return false
    }
    ok := false
    for !ok {
      fmt.Printf("WHAT DO YOU WANT TO DO? (%v)\n",list)
      key, err := reader.ReadString('\n')
      if err != nil {
        panic(err)
      }
      action, ok = combatActions[key]
      if !ok {
        fmt.Printf("THIS IS NOT A VALID ACTION (%v)\n", key)
      }
    }
    repeat = action(player, monster)
  }
}

var combatActions = map[string]func(player *Player, monster *Monster) bool {
  "ATTACK":func(player *Player, monster *Monster) bool {
    if rand.Intn(2) > 1 {

    } else {

    }
    return false
  },
  "FLEE":func(player *Player, monster *Monster) bool {
    if rand.Intn(10) < 7 {
      fmt.Println("YOU RUN AWAY!")
      last := len(player.Path) - 1
      player.CurrentRoom = player.Path[last]
      player.Path = player.Path[:last]
      return false
    } else {
      fmt.Println("YOU CANNOT RUN AWAY!")
      fmt.Println("YOU MUST STAND AND FIGHT")
      return true
    }
  },
  "EAT":func(player *Player, monster *Monster) bool {
    item, ok := player.Inventory["FOOD"]
    if !ok {
      fmt.Println("You have no food")
    }
    if item.Count <= 0 {
      fmt.Println("You have no food")
    }
    player.Strength += 5
    item.Count -= 1
    return true;
  },
}

func buildGame(xml GameXML, name string) *Game {
  game := &Game{}
  game.Player = &Player{}
  game.Player.Name = name
  game.Player.CurrentRoom = xml.Init.FirstRoom
  game.Player.Path = []int{}
  game.Player.Strength = xml.Init.InitStats.Strength
  game.Player.Wealth = xml.Init.InitStats.Wealth
  game.Player.Tally = xml.Init.InitStats.Tally
  game.Player.Inventory = map[string]*InventoryItem{}
  game.Maze.Entrance = xml.Init.FirstRoom
  game.Maze.Exit = xml.Init.LastRoom
  game.Maze.Rooms = map[int]*Room{}
  for _, roomXML := range xml.Rooms {
    room := &Room{}
    room.Description = roomXML.Lines
    room.Contents = roomXML.Treasure.Eval()
    room.Doors = map[string]int{}
    if roomXML.Doors.North < 0 {
      room.Doors["NORTH"] = roomXML.Doors.North
    }
    if roomXML.Doors.South < 0 {
      room.Doors["SOUTH"] = roomXML.Doors.South
    }
    if roomXML.Doors.East < 0 {
      room.Doors["EAST"] = roomXML.Doors.East
    }
    if roomXML.Doors.West < 0 {
      room.Doors["WEST"] = roomXML.Doors.West
    }
    if roomXML.Doors.Up < 0 {
      room.Doors["UP"] = roomXML.Doors.Up
    }
    if roomXML.Doors.Down < 0 {
      room.Doors["DOWN"] = roomXML.Doors.Down
    }
    game.Maze.Rooms[roomXML.Index] = room
  }
  game.Monsters = map[int]*Monster{}
  for _, monsterXml := range xml.Monsters {
    monster := &Monster{}
    monster.Name = monsterXml.Name
    monster.Danger = monsterXml.DangerLevel
    game.Monsters[monsterXml.Index] = monster
  }
  game.Shop = map[string]*Item{}
  for _, shopItemXml := range xml.ShopItems {
    item := &Item{}
    item.Name = shopItemXml.Name
    item.Cost = shopItemXml.Cost
    item.Limit = shopItemXml.Limit
    item.Effect = shopItemXml.Effect
    game.Shop[shopItemXml.Name] = item
  }
  rooms := []int{}
  for index, room := range game.Maze.Rooms {
    if index != game.Maze.Entrance && index != game.Maze.Exit && room.Contents == 0 {
      rooms = append(rooms, index);
    }
  }
  for monsterIndex, _ := range game.Monsters {
    room := &Room{}
    ok := false
    for !ok {
      roomNumber := rand.Intn(len(rooms))
      room, ok = game.Maze.Rooms[roomNumber]
      room.Contents = monsterIndex
      rooms = append(rooms[:roomNumber], rooms[roomNumber+1:]...)
    }
  }
  placeTreasure := xml.Init.PlaceTreasure
  for count := 0; count < placeTreasure.RoomCount; count++ {
    room := &Room{}
    ok := false
    for !ok {
      roomNumber := rand.Intn(len(rooms))
      room, ok = game.Maze.Rooms[roomNumber]
      room.Contents = placeTreasure.Treasure.Eval()
      rooms = append(rooms[:roomNumber], rooms[roomNumber+1:]...)
    }
  }
  sort.Ints(rooms)
  return game
}

File: werewolves/game.js
var gamefactory = (function(game){
  var state = util.initgame(game)
  return {
    build:function(out, err, menu, optionfactory) {
      var logs = {
        readout:[],
        errors:[]
      }
      writeout = util.writer(out, logs, "readout")
      writeerr = util.writer(err, logs, "errors")
      var io = {
        in:function(text) {
          writeout(text)
          var input = prompt(text.join("\n"))
          if (input == null) {
            throw "io escaped"
          }
          writeout(input)
          return input;
        },
        out:function(text) {
          writeout(text, true)
        },
        err:function(text) {
          writeerr(text, true)
        }
      }
      var gameactions = util.gameactionsfactory(state, io)
      var travelActions = {
        "MOVE":function(self) {
          var direction = inquireOfUser(
            "WHERE DO YOU WANT TO GO?",
            state.game.rooms[state.player.room].doors
          )
          gameactions.MOVE(direction)
        },
        "BUY":function(self) {
          if (self.wealth <= 0) {
            io.out([
              "YOU DO NOT HAVE ANY GOLD",
              "YOU CANNOT BUY ANYTHING"
            ])
          } else {
            io.out(state.player.readout())
            io.out(state.game.storelisting())
            var itemname = inquireOfUser("WHAT DO YOU WANT TO BUY?", state.game.store)
            gameactions.PURCHASE(itemname)
          }
        },
        "EAT":gameactions.EAT,
        "EQUIP":function(self) {
          state.player.equiptment = gameactions.LIST_EQUIPTMENT();
          if (Object.keys(equiptment).length == 0) {
            io.out("YOU HAVE NOTHING TO EQUIP")
          } else {
            var item = inquireOfUser("WHAT DO YOU WANT TO EQUIP?", equiptment)
            gameactions.EQUIP_SELECTED(equiptment,item)
          }
        },
        "STATUS":gameactions.STATUS
      }
      var combatActions = {
        "FIGHT":gameactions.FIGHT,
        "RUN":gameactions.RUN,
        "EAT":gameactions.EAT
      }
      function inquireOfUser(question, options) {
        var list = "(" + Object.keys(options).join(",").toUpperCase() + ")"
        message = [question, list];
        while (true) {
          var answer = io.in(message)
          if (answer in options) {
            return answer
          }
          answer = answer.toUpperCase()
          if (answer in options) {
            return answer
          }
          answer = answer.toLowerCase()
          if (answer in options) {
            return answer
          }
        }
      }
      function populateSelector(selector, optionfactory, question, options) {
        var option = optionfactory();
        option.text = question;
        option.value = "";
        selector.add(option);
        Object.keys(options).forEach(function(key) {
          var option = optionfactory();
          option.text = key;
          option.value = key;
          selector.add(option);
        })
        selector.enabled
      }
      function retrieveValue(selector) {
        var selected = selector.value
        if (selected.length > 0) {
          selector.disabled = true
          while (selector.length > 0) {
            selector.remove(0)
          }
          return selected
        }
      }
      var travelEvents = {
        "MOVE":function(self) {
          state.state = "MOVE"
          populateSelector(
            menu,
            optionfactory,
            "WHERE DO YOU WANT TO GO?",
            state.game.rooms[state.player.room].doors
          )
        },
        "BUY":function(self) {
          if (self.wealth <= 0) {
            io.out([
              "YOU DO NOT HAVE ANY GOLD",
              "YOU CANNOT BUY ANYTHING"
            ])
            delete state.state
            events.MAIN()
          } else {
            io.out(state.player.readout())
            io.out(state.game.storelisting())
            state.state = "PURCHASE"
            populateSelector("WHAT DO YOU WANT TO BUY?", state.game.store)
        },
        "EAT":gameactions.EAT,
        "EQUIP":function(self) {
          state.player.equiptment = gameactions.LIST_EQUIPTMENT();
          if (Object.keys(equiptment).length == 0) {
            io.out("YOU HAVE NOTHING TO EQUIP")
          } else {
            state.state = "EQUIP"
            populateSelector("WHAT DO YOU WANT TO EQUIP?", equiptment)
          }
        },
        "STATUS":gameactions.STATUS
      }
      var combatEvents = {
        "FIGHT":function(){
          var turns = {
            you:function() {
              io.out("YOU ATTACK")
              if (Math.random() > .5) {
                io.out("YOU MANAGE TO WOUND IT")
                target.danger = Math.floor(5 * target.danger / 6)
              } else {
                io.out("YOU MISS")
              }
            },
            him:function() {
              io.out("THE " + target.name + " ATTACKS")
              if (Math.random() > .5) {
                io.out("THR MONSTER WOUNDS YOU")
                self.strength -= 5
                io.out(self.readout())
                return self.strength <= 0
              } else {
                io.out("HE MISSED")
              }
            }
          }
          var initiative = []
          if (Math.random() > .5) {
            initiative.push("him")
            initiative.push("you")
          } else {
            initiative.push("him")
            initiative.push("you")
          }
          for (var x = 0; x < initiative.length; x++) {
            if (turns[initiative[x]]()) {
              break;
            }
          }
          if (self.strength <= 0) {
            return true
          }
          if (Math.random() <= 0.35) {
            return true;
          }
        },
        "RUN":function(){
          if (Math.random() > .7) {
            io.out("NO, YOU MUST STAND AND FIGHT")
            // TODO --
          } else {
            self.room = self.path.pop()
            io.out("YOU RUN BACK TO THE ROOM YOU WERE IN LAST")
            delete state.state
            events.INIT()
          }
        },
        "EAT":gameactions.EAT
      }
      var events = {
        INIT:function() {
          io.out(state.player.readout())
          events.MAIN()
        },
        MAIN:function() {
          io.out(state.game.rooms[state.player.room].description)
          if (state.player.strength == 0) {
            io.out("YOU ARE DEAD.....")
          } else if (state.player.room != state.game.init.exit) {
            var contents = state.game.rooms[state.player.room].contents
            if (contents < 0) {
              var monster = state.game.monsters[contents]
              var target = gameactions.INIT_COMBAT(state.player, monster)
              state.state = "COMBAT"
              populateSelector(menu, optionfactory, "WHAT DO YOU WANT TO DO?", combatActions)
            } else {
              if (contents > 0) {
                gameactions.ACQUIRE(state.player, contents)
              }
              state.state = "TRAVEL"
              populateSelector(menu, optionfactory, "WHAT DO YOU WANT TO DO?", travelEvents)
            }
          }
        },
        TRAVEL:function(opt) {
          travelEvents[opt](state.player)
        },
        MOVE:function(opt) {
          delete state.state
          gameactions.MOVE(direction)
          events.MAIN()
        },
        COMBAT:function(opt) {
          var contents = state.game.rooms[state.player.room].contents
          if (contents < 0) {
            var monster = state.game.monsters[contents]
            combatActions[combatAction](state.player, monster)
          }
          // TODO --
        },
        PURCHASE:function(opt) {
          delete state.state
          gameactions.PURCHASE(opt)
          events.MAIN()
        },
        EQUIP:function(opt) {
          delete state.state
          gameactions.EQUIP_SELECTED(state.player.equiptment,item)
          delete state.player.equiptment
        }
      };
      return {
        play:function() {
          io.out(state.player.readout())
          while (state.player.room != state.game.init.exit) {
            io.out(state.game.rooms[state.player.room].description)
            var contents = state.game.rooms[state.player.room].contents
            if (contents > 0) {
              gameactions.ACQUIRE(state.player, contents)
            } else if (contents < 0) {
              var monster = state.game.monsters[contents]
              var target = gameactions.INIT_COMBAT(state.player, monster)
              var concluded = false
              var runaway = false
              while (!concluded) {
                try {
                  var combatAction = inquireOfUser("WHAT DO YOU WANT TO DO?", combatActions)
                  concluded = combatActions[combatAction](state.player, target)
                } catch (e) {
                  if (e == "RUN") {
                    runaway = true
                    break;
                  }
                  throw e;
                }
              }
              if (!runaway) {
                gameactions.CONCLUDE_COMBAT(state.player, monster)
              }
              if (state.player.strength == 0) {
                break;
              }
            }
            var action = inquireOfUser("WHAT DO YOU WANT TO DO?", travelActions)
            travelActions[action](state.player)
          }
          gameactions.CONCLUDE_GAME()
        },
        resolve:function() {
          var value = retrieveValue(menu);
          if (value) {
            events[state.state](menu.value)
          }
        }
      }
    }
  }
})(were)
